plone/app/vocabularies/principals.py

```diff
@@ -0,0 +1,203 @@
+        """
+        try:
+            return super(PrincipalsVocabulary, self).getTerm(value)
+        except LookupError:
+            user = self._get_from_source(value)
+            if user is None:
+                raise
+            if user.isGroup():
+                title = user.getProperty('title', value) or value
+            else:
+                title = user.getProperty('fullname', value) or value
+            if SOURCES[self._principal_source]['prefix']:
+                token = value.replace(':', '__', 1)
+            else:
+                token = value
+            return self.__class__.createTerm(value, token, title)
+
+    def __getitem__(self, start, stop=None):
+        """Sliceable"""
+        if isinstance(start, slice):
+            slice_inst = start
+            start = slice_inst.start
+            stop = slice_inst.stop
+        elif not stop:
+            return self._terms[start]
+
+        # sliced up
+        return self._terms[start:stop]
+
+
+class BaseFactory(object):
+    """Factory creating a PrincipalsVocabulary
+    """
+
+    source = None
+
+    def should_search(self, query):
+        """ Test if we should search for users
+        """
+        if query:
+            return True
+        registry = getUtility(IRegistry)
+        return not [
+            x
+            for x in filter(
+                registry.get,
+                [cfg['many'] for cfg in SOURCES[self.source]['searches']],
+            )
+        ]
+
+    def use_principal_triple(self, principal_triple):
+        """Used by ``functools.filter`` to decide if the triple should be used.
+
+        principal_triple
+            A triple (token, value, title).
+            Like (johndoe, johndoe, 'John Doe') (unprefixed).
+            Value might be a prefixed Id by principal_type, like
+            (user__johndoe, user:johndoe, 'John Doe') or
+            (group__editors, group:editors, 'Editors').
+
+        returns wether the triple shall be included in the vocabulary or not
+        (bool).
+
+        Meant to be overriden in subclasses for post-filtering result.
+        """
+        return True
+
+    def __call__(self, context, query=''):
+        if not self.should_search(query):
+            return PrincipalsVocabulary.fromItems([])
+        acl_users = _get_acl_users()
+        cfg = SOURCES[self.source]
+
+        def principal_triples():
+            for search_cfg in cfg['searches']:
+                search = getattr(acl_users, search_cfg['search'])
+                searchargs = search_cfg['searchargs'].copy()
+                searchargs[search_cfg['searchattr']] = query
+                for info in search(**searchargs):
+                    principal_id = info['id']
+                    if cfg['prefix']:
+                        principal_id = '{0}:{1}'.format(
+                            info['principal_type'], principal_id
+                        )
+                    principal_token = principal_id.replace(':', '__')
```

Comment:
The principals-vocabulary is broken when I use the pricipals-vocabulary in a type like below. Selecting and searching works fine but when I save this happens. Notice that also the field without validation has the token with prefix instead of the title. 

<img width="496" alt="principals" src="https://user-images.githubusercontent.com/453208/54216173-6ee3c380-44e9-11e9-9533-3c2ba4ac8625.png">

```python 
# -*- coding: utf-8 -*-
from plone.app.z3cform.widget import AjaxSelectFieldWidget
from plone.autoform import directives
from plone.dexterity.content import Container
from plone.supermodel import model
from zope import schema
from zope.interface import implementer


class IDummy(model.Schema):
    """Dexterity-Schema for contracts
    """

    users = schema.Tuple(
        title=u'Users',
        value_type=schema.TextLine(),
        required=False,
        missing_value=(),
    )
    directives.widget(
        'users',
        AjaxSelectFieldWidget,
        vocabulary='plone.app.vocabularies.Users',
        pattern_options={'allowNewItems': False},
    )

    groups = schema.Tuple(
        title=u'Groups',
        value_type=schema.TextLine(),
        required=False,
        missing_value=(),
    )
    directives.widget(
        'groups',
        AjaxSelectFieldWidget,
        vocabulary='plone.app.vocabularies.Groups',
        pattern_options={'allowNewItems': False},
    )

    principals = schema.Tuple(
        title=u'Principals',
        value_type=schema.TextLine(),
        required=False,
        missing_value=(),
    )
    directives.widget(
        'principals',
        AjaxSelectFieldWidget,
        vocabulary='kbo.webapp.vocabularies.Principals',
        pattern_options={'allowNewItems': False},
    )

    users_valid = schema.Tuple(
        title=u'Users validated',
        value_type=schema.Choice(
            vocabulary='plone.app.vocabularies.Users'),
        required=False,
        missing_value=(),
    )
    directives.widget(
        'users_valid',
        AjaxSelectFieldWidget,
        vocabulary='plone.app.vocabularies.Users',
        pattern_options={'allowNewItems': False},
    )

    groups_valid = schema.Tuple(
        title=u'Groups validated',
        value_type=schema.Choice(
            vocabulary='plone.app.vocabularies.Groups'),
        required=False,
        missing_value=(),
    )
    directives.widget(
        'groups_valid',
        AjaxSelectFieldWidget,
        vocabulary='plone.app.vocabularies.Groups',
        pattern_options={'allowNewItems': False},
    )

    principals_valid = schema.Tuple(
        title=u'Principals validated',
        value_type=schema.Choice(
            vocabulary='kbo.webapp.vocabularies.Principals'),
        required=False,
        missing_value=(),
    )
    directives.widget(
        'principals_valid',
        AjaxSelectFieldWidget,
        vocabulary='kbo.webapp.vocabularies.Principals',
        pattern_options={'allowNewItems': False},
    )


@implementer(IDummy)
class Dummy(Container):
    """Contenttype class.
    """
```

Reply:
My mistake. I used a broken fix for the `ValueError: too many values to unpack` issue.

Reply:
Validation works now with https://github.com/plone/plone.app.vocabularies/pull/57 but when editing the prefixed value is displayed instead of the title:

<img width="757" alt="principals_edit" src="https://user-images.githubusercontent.com/453208/54220688-3b596700-44f2-11e9-891c-d6cb268f0026.png">

Reply:
The reason is that `initialValues` has the token (e.g. `group__appm_pflege`) and the value has the value (`group:appm_pflege`). I guess select2 expect the two to be the same (or somehow similar) to be able to match the values and pick the correct title to display.

This is how the widget looks currently with this vocabulary:

```html
<input 
    class="pat-select2 text-widget list-field" 
    type="text" 
    name="form.widgets.informed" 
    value="group:appm_pflege;user:appm_pflege_member" 
    data-pat-select2="{'allowNewItems': 'false', 
                       'orderable': true, 
                       'vocabularyUrl': 'http://localhost:8080/Plone/massnahmen/m20190001/@@getVocabulary?name=kbo.webapp.vocabularies.Principals&field=informed', 
                       'initialValues': {'group__appm_pflege': 'APPM Pflege', 'user__appm_pflege_member': 'APPM Pflege Member'}, 'separator': ';'}">
```

When I change the vocabulary that for the token I do not replace the `:` with `__` it looks like this and works (at least in Firefox):

```html
<input 
    class="pat-select2 text-widget list-field" 
    type="text" 
    name="form.widgets.informed" 
    value="group:appm_pflege;user:appm_pflege_member" 
    data-pat-select2="{'allowNewItems': 'false', 
                       'orderable': true, 
                       'vocabularyUrl': 'http://localhost:8080/Plone/massnahmen/m20190001/@@getVocabulary?name=kbo.webapp.vocabularies.Principals&field=informed', 
                       'initialValues': {'group:appm_pflege': 'APPM Pflege', 'user:appm_pflege_member': 'APPM Pflege Member'}, 
                       'separator': ';'}">

```

Is the reason you replace the `:` with `__` browser-support?

Reply:
More or less: Yes. I had in mind that using `:` in token was a bad idea for some reason, seems here it is. I think I have a idea to how to keep the `__` and make it work, just give me some minutes...

Reply:
I think `:` is problematic because it is the default-separator when passing `initialValues` to `select2` as a string (not as json like in my example above):
https://github.com/plone/mockup/blob/master/mockup/patterns/select2/pattern.js#L106

Reply:
Did you find a solution that worked?